controllers/provider-openstack/pkg/controller/infrastructure/actuator_reconcile.go

```diff
@@ -0,0 +1,140 @@
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/chartrenderer"
+	"github.com/gardener/gardener/pkg/operation/terraformer"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	defaultRouterID = "${openstack_networking_router_v2.router.id}"
+	// DomainName is a constant for the key in a cloud provider secret that holds the OpenStack domain name.
+	DomainName = "domainName"
+	// TenantName is a constant for the key in a cloud provider secret that holds the OpenStack tenant name.
+	TenantName = "tenantName"
+	// UserName is a constant for the key in a cloud provider secret and backup secret that holds the OpenStack username.
+	UserName = "username"
+	// Password is a constant for the key in a cloud provider secret and backup secret that holds the OpenStack password.
+	Password = "password"
+)
+
+type credentials struct {
+	DomainName string
+	TenantName string
+}
+
+func (a *actuator) reconcile(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, cluster *extensionscontroller.Cluster) error {
+	infrastructureConfig := &openstackv1alpha1.InfrastructureConfig{}
+	if _, _, err := a.decoder.Decode(infrastructure.Spec.ProviderConfig.Raw, nil, infrastructureConfig); err != nil {
+		return fmt.Errorf("could not decode provider config: %+v", err)
+	}
+
+	providerSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, kutil.Key(infrastructure.Spec.SecretRef.Namespace, infrastructure.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	terraformConfig, err := generateTerraformInfraConfig(ctx, infrastructure, infrastructureConfig, cluster, extractCredentials(providerSecret))
+	if err != nil {
+		return fmt.Errorf("failed to generate Terraform config: %+v", err)
+	}
+
+	chartRenderer, err := chartrenderer.NewForConfig(a.restConfig)
+	if err != nil {
+		return fmt.Errorf("could not create chart renderer: %+v", err)
+	}
+
+	release, err := chartRenderer.Render(filepath.Join(openstack.InternalChartsPath, "openstack-infra"), "openstack-infra", infrastructure.Namespace, terraformConfig)
+	if err != nil {
+		return fmt.Errorf("could not render Terraform chart: %+v", err)
+	}
+
+	tf, err := a.newTerraformer(openstack.TerrformerPurposeInfra, infrastructure.Namespace, infrastructure.Name)
+	if err != nil {
+		return fmt.Errorf("could not create terraformer object: %+v", err)
+	}
+
+	if err := tf.
+		SetVariablesEnvironment(generateTerraformInfraVariablesEnvironment(providerSecret)).
+		InitializeWith(terraformer.DefaultInitializer(
+			a.client,
+			release.FileContent("main.tf"),
+			release.FileContent("variables.tf"),
+			[]byte(release.FileContent("terraform.tfvars"))),
+		).
+		Apply(); err != nil {
+
+		return &controllererrors.RequeueAfterError{
+			Cause:        err,
+			RequeueAfter: 30 * time.Second,
+		}
+	}
+
+	return nil
```

Comment:
You should generate the `InfrastructureStatus` here, see https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-aws/pkg/controller/infrastructure/actuator_reconcile.go#L90-L92

Reply:
Good point. Totally forgot about the status.

Reply:
Done.